### PR TITLE
[Network] Fix the windows ping problem for local subnets.

### PR DESCRIPTION
**Pull request by 9037568**
Related to #3502.

Fixes the problem described in [this community thread](https://community.openhab.org/t/windows-ping-does-not-work-correct/43776/24) where Windows pings from the Network binding to hosts on the local subnet are incorrectly parsed, leading to unreachable hosts being flagged as reachable.

This is the [test jar](https://drive.google.com/open?id=17CLqpE7mWSLggY8rtMW3SNNzkyiJDon5).

Signed-off-by: 9037568 <namraccr@gmail.com>

### CODE REVIEW

File: addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java [hunk 1]
```diff
@@ -254,9 +260,42 @@ public boolean nativePing(IpPingMethodEnum method, String hostname, int timeoutI
...
+            return proc.waitFor() == 0;
+        }
+
+        int result = proc.waitFor();
+        if (result != 0) {
+            return false;
+        }
+
+        try{
+            BufferedReader r = new BufferedReader(new InputStreamReader(proc.getInputStream()));
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: Shouldn't you better use try-with-resources here?

File: addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java [hunk 2]
```diff
@@ -32,13 +34,17 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Network utility functions for pinging and for determining all interfaces and assigned IP addresses.
  *
  * @author David Graeff <david.graeff@web.de>
  */
 public class NetworkUtils {
+    private Logger logger = LoggerFactory.getLogger(NetworkUtils.class);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: please add `final`.

File: addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java [hunk 3]
```diff
@@ -254,9 +260,42 @@ public boolean nativePing(IpPingMethodEnum method, String hostname, int timeoutI
...
+            }
+            do {
+                logger.trace("Examining ping process line: '{}'", line);
+                if (line.contains("host unreachable") ||
+                    line.contains("timed out") ||
+                    line.contains("could not find host")) {
+                    return false;
+                }
+                line = r.readLine();
+            }while(line != null);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: You do not seem to use the openHAB IDE with the automatic code formatter?

File: addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java (hunk 2 above)
Comment (9037568) [reply addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: Done.

File: addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java (hunk 3 above)
Comment (9037568) [reply addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: I usually remember to reformat before committing. Will fix this.

File: addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java (hunk 1 above)
Comment (9037568) [reply addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: Will do. I don't need to use that construct very often.

File: addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java [hunk 4]
```diff
@@ -254,9 +260,40 @@ public boolean nativePing(IpPingMethodEnum method, String hostname, int timeoutI
...
+        }
+
+        try (BufferedReader r = new BufferedReader(new InputStreamReader(proc.getInputStream()))) {
+            String line = r.readLine();
+            if (line == null) {
+                logger.trace("Received no output from ping process.");
+                return false;
+            }
+            do {
+                logger.trace("Examining ping process line: '{}'", line);
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: Personally I would remove this logging line. Actually I tried to keep this utility class logger free on purpose. But if we agree on introducing a logger here, it should at least not be used in this hot code-path for a well-known output. For a local subnet scan for example, this method is called in parallel for every single valid IP in range and the logger is not a no-op.

File: addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java [hunk 5]
```diff
@@ -254,9 +260,40 @@ public boolean nativePing(IpPingMethodEnum method, String hostname, int timeoutI
...
+
+        int result = proc.waitFor();
+        if (result != 0) {
+            return false;
+        }
+
+        try (BufferedReader r = new BufferedReader(new InputStreamReader(proc.getInputStream()))) {
+            String line = r.readLine();
+            if (line == null) {
+                logger.trace("Received no output from ping process.");
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: This should never happen, should it? I'd say the logger level might be raised to a warning, or alternatively throw an IOException.

Comment (9037568) [reply addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: I removed the logging and put in throws.

File: addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java [hunk 6]
```diff
@@ -254,9 +256,37 @@ public boolean nativePing(IpPingMethodEnum method, String hostname, int timeoutI
...
+            do {
+                if (line.contains("host unreachable") || line.contains("timed out")
+                        || line.contains("could not find host")) {
+                    return false;
+                }
+                line = r.readLine();
+            } while (line != null);
+
+            return true;
+        } catch (IOException e) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: You shouldjust leave away this catch/throw when using try-with-resources

Comment (martinvw) [reply addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java]: Never mind I did it myself, will wait for the build to be 101% sure

**Comment by martinvw** [comment]
Thanks!

**Pull request merged**